content/en/blog/2022/rust-instrumentation.md

```diff
@@ -0,0 +1,378 @@
+---
+title: Rust & Otel Instrumentation
+linkTitle: Rust & Otel Instrumentation
+date: 2022-09-21
+author: '[Yoav Danieli](https://github.com/aspecto-io) (Aspecto)'
+canonical_url: https://www.aspecto.io/blog/distributed-tracing-with-opentelemetry-rust/
+---
+
+# Getting Started: OpenTelemetry in Rust
+
+![](https://lh3.googleusercontent.com/vlVY92XQhT69sNU4ZukikMN6fBpftvvf3Qs1zizIO1jJiwhB3giIpFE4ZM1q09JMdDQvl8Z7ceBFh4sebPYlrrDsdstJUILalWw3IW6OMAci2bwusf5DUtILclDYSg3zYGsUXhrlCvPzPtvmZx2UGgWj79rRFUCWDzzAvaQeBMEjlv0swZETT-XEiQ)
+
+In this article, I will share my experience adding OpenTelemetry distributed tracing to a Rust application.
+
+I will attempt to answer the following questions:
+
+- How to instrument Opentelemetry in Rust?
+
+- How to add manual and auto instrumentations to a Rust application?
+
+- How to use tracing to debug Rust applications?
+
+- How to visualize traces and spans?
+
+- How to preserve our span context in a multithreaded environment?
+
+- What are your must-know crates when it comes to tracing in Rust?
+
+I will begin by saying that I love Rust as a programming language, and I've been experimenting with it for the past year. Although it's innovative and has a great ecosystem, it is not a secret that Rust has quite a steep learning curve. So even though it's an intro-level article, I will not dive deep into Rust features, syntax, or semantics unless it's somehow connected to our subject (Opentelemetry and distributed tracing).
+
+## Opentelemetry and Distributed Tracing in Rust
+
+### Tracing in Rust
+
+In Rust we have a great crate simply called “[tracing](https://docs.rs/tracing/latest/tracing/)”. It is our core framework for tracing a Rust program. The crates documentation outlines its core concepts and APIs which are composed of these terms - spans, events, and subscribers.
+
+I covered what Spans and Events are, but a Subscriber is a new term. To record spans and events, one has to implement the subscriber trait. That means simply implementing the methods:
```

Comment:
You didn't actually cover what spans and events are here, right?

The wording here seems unclear to me: "to record spans and events, one has to implement the subscriber trait". I would argue that recording and subscribing are different things, happening at different times, and this sentence doesn't do a great job explaining the concept.

Also, most users will not implement their own subscriber, instead initializing existing subscribers, so the methods to implement here seem irrelevant to most folks.